src/C4EnergyBars.cpp

```diff
@@ -0,0 +1,586 @@
+{
+	C4Value name = bar[C4VString("name")];
+	auto *_name = name.getStr();
+	if (!_name) {
+		LogF("DefineEnergyBars: EnergyBar definition has no name: %s", name.GetDataString().getData());
+		return false;
+	}
+
+	C4Value gfx = bar[C4VString("gfx")];
+	auto *_gfx = gfx.getStr();
+	if (!_gfx) {
+		LogF("DefineEnergyBars: EnergyBar definition has no gfx: %s", gfx.GetDataString().getData());
+		return false;
+	}
+
+	C4Value physical = bar[C4VString("physical")];
+	C4ValueInt _physical = physical.getInt();
+	if (_physical & ~C4EnergyBarDef::EBP_All) {
+		LogF("DefineEnergyBars %s definition has invalid physical: %s", name.GetDataString().getData(), physical.GetDataString().getData());
+		return false;
+	}
+
+	C4Value hide = bar[C4VString("hide")];
+	C4ValueInt _hide = C4EnergyBarDef::EBH_Empty;
+	if (hide != C4VNull) _hide = hide.getInt();
+	if (_hide & ~C4EnergyBarDef::EBH_All) {
+		LogF("DefineEnergyBars %s definition has invalid hide: %s", name.GetDataString().getData(), hide.GetDataString().getData());
+		return false;
+	}
+
+	C4Value index = bar[C4VString("index")];
+	C4Value value = bar[C4VString("value")];
+	C4Value max   = bar[C4VString("max")];
+	C4ValueInt _index = index.getInt();
+	C4ValueInt _value = value.getInt();
+	C4ValueInt _max   = max.getInt();
+	if (_index < 0) {LogF("DefineEnergyBars %s definition has invalid index: %s", name.GetDataString().getData(), _index); return false;}
+	if (_value < 0) {LogF("DefineEnergyBars %s definition has invalid value: %s", name.GetDataString().getData(), _value); return false;}
+	if (_max   < 0) {LogF("DefineEnergyBars %s definition has invalid max: %s", name.GetDataString().getData(), _max); return false;}
+
+	bool _visible = true;
+	if (bar.contains(C4VString("visible"))) {
+		auto visible = bar[C4VString("visible")];
+		_visible = visible.getBool();
+	}
+
+	{
+		const char* file = _gfx->Data.getData();
+		auto facet = GetFacet(graphics, file);
+		if(facet == nullptr) {return false;}
+
+		C4EnergyBarDef bar(_name->Data.getData(), file, facet, _index, _physical);
+		if(physical != C4VNull) {
+			bar.physical = _physical;
+			bar.hide = C4EnergyBarDef::DefaultHide(_physical);
+		} else {
+			bar.value_index = value_index++;
+		}
+		if(hide != C4VNull) bar.hide = _hide;
+		bar.value = _value;
+		bar.max = _max;
+		bar.visible = _visible;
+		bar.advance = advance;
+		auto scale = graphics.at(file).scale;
+		bar.scale = static_cast<float>(scale) / 100.0f;
+		bars.push_back(bar);
+	}
+
+	return true;
+}
+
+std::shared_ptr<C4EnergyBars> C4EnergyBarsUniquifier::Instantiate(std::shared_ptr<C4EnergyBarsDef> definition)
+{
+	if(definition == nullptr) return nullptr;
+	return std::make_shared<C4EnergyBars>(definition);
+}
+
+void C4EnergyBarsAdapt::CompileFunc(StdCompiler *pComp) {
+	bool fCompiler = pComp->isCompiler();
+
+	if (!fCompiler) {
+		// serialize
```

Comment:
this and deserialize can be assumed self-explanatory in this code base, imo

Reply:
Ok, personally I find the isCompiler-Naming confusing, never quite know whether it is reading/writing or deserialising/serialising. Having to look back into the implementation of StdCompiler. Maybe it gets better with time, or maybe I find a way to remember which way it is...

Reply:
I get that.
Personally, I always think about what compiling means.
For me it means reading something human readable (thinking about Clonks INI format) and converting it to program data.
And the opposite way would be converting program data back to the readable format. Maybe that helps somehow.